http/router/router.go

```diff
@@ -1,95 +1,143 @@
+		*wrapper = *r
+		wrapper.chi = c
+		fn(wrapper)
+	})
 }
-func (r *Router) Put(pattern string, fn handlerFunc) {
-	r.Chi.Put(pattern, handler(fn))
+
+// Method adds a routes for a `pattern` that matches the `method` HTTP method.
+func (r *chiRouter) Method(method, pattern string, h HandlerErrorFunc) {
+	f := HandlerFunc(h)
+	r.chi.Method(method, pattern, f)
 }
-func (r *Router) Delete(pattern string, fn handlerFunc) {
-	r.Chi.Delete(pattern, handler(fn))
+
+// Get adds a GET route
+func (r *chiRouter) Get(pattern string, fn HandlerErrorFunc) {
+	r.chi.Get(pattern, HandlerFunc(fn))
 }
 
-func (r *Router) With(fn MiddlewareHandler) *Router {
-	c := r.Chi.With(middleware(fn))
-	return &Router{c}
+// Post adds a POST route
+func (r *chiRouter) Post(pattern string, fn HandlerErrorFunc) {
+	r.chi.Post(pattern, HandlerFunc(fn))
 }
 
-func (r *Router) Use(fn MiddlewareHandler) {
-	r.Chi.Use(middleware(fn))
+// Put adds a PUT route
+func (r *chiRouter) Put(pattern string, fn HandlerErrorFunc) {
+	r.chi.Put(pattern, HandlerFunc(fn))
 }
-func (r *Router) UseBypass(fn func(next http.Handler) http.Handler) {
-	r.Chi.Use(fn)
+
+// Delete adds a DELETE route
+func (r *chiRouter) Delete(pattern string, fn HandlerErrorFunc) {
+	r.chi.Delete(pattern, HandlerFunc(fn))
 }
 
-func (r *Router) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	r.Chi.ServeHTTP(w, req)
+// With adds an inline chi middleware for an endpoint handler
+func (r *chiRouter) With(fn Middleware) Router {
+	r.chi = r.chi.With(fn)
+	return r
 }
 
-type handlerFunc func(w http.ResponseWriter, r *http.Request) error
+// Use appends one chi middleware onto the Router stack
+func (r *chiRouter) Use(fn Middleware) {
+	r.chi.Use(fn)
+}
 
-func handler(fn handlerFunc) http.HandlerFunc {
-	return fn.serve
+// ServeHTTP will serve a request
+func (r *chiRouter) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	r.chi.ServeHTTP(w, req)
 }
 
-func (h handlerFunc) serve(w http.ResponseWriter, r *http.Request) {
-	if err := h(w, r); err != nil {
-		httperr.HandleError(err, w, r)
-	}
+// Mount attaches another http.Handler along ./pattern/*
+func (r *chiRouter) Mount(pattern string, h http.Handler) {
+	r.chi.Mount(pattern, h)
 }
 
-type MiddlewareHandler func(w http.ResponseWriter, r *http.Request) (context.Context, error)
+// =======================================
+// HTTP handler with custom error payload
+// =======================================
 
-func (m MiddlewareHandler) handler(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		m.serve(next, w, r)
-	})
-}
+type HandlerErrorFunc func(w http.ResponseWriter, r *http.Request) error
 
-func (m MiddlewareHandler) serve(next http.Handler, w http.ResponseWriter, r *http.Request) {
-	ctx, err := m(w, r)
-	if err != nil {
-		httperr.HandleError(err, w, r)
-		return
-	}
-	if ctx != nil {
-		r = r.WithContext(ctx)
+func HandlerFunc(fn HandlerErrorFunc) http.HandlerFunc {
```

Comment:
The `HandlerFunc` naming here is very confusing due to conflicting name with a similar type in the "http" package.

I would add a method to the HandlerErrorFunc type named ServeHTTP, thus our HandlerErrorFunc type will also implement the http.Handler interface and it will also keep the `router` package API clean:

```
func(f HandlerErrorFunc) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if err := f(w, r); err != nil {
		commonshttp.HandleError(err, w, r)
	}
}
```